### PR TITLE
feat(processing): add Smooth, Regular grid, Voronoi/Delaunay vector tools

### PR DESCRIPTION
**Pull request by giswqs**
## Summary
- Adds the three remaining vector tools tracked in #283: Smooth (Chaikin corner-cutting, distinct from Simplify), Regular grid (fishnet of rectangles or cell-center points), and Voronoi / Delaunay from a point layer. Each is wired into the Processing -> Vector menu and the command palette.
- Smooth and Voronoi/Delaunay run on both the client (Turf.js) and the Python sidecar/Pyodide engines; Smooth shares one Chaikin implementation across engines so results are identical, and Voronoi clips cells to the points' bounding box plus a 10% margin on both. Regular grid is client-side (like the H3 tools) with a hard cell cap.
- Confirms the related #283 cleanup item is already resolved: the dead registry placeholders are gone and the format placeholder messages were refreshed in an earlier change; the only remaining placeholder message covers a genuinely unshipped feature.

## Test plan
- [x] Frontend `node:test` covers the Smooth, Regular grid, and Voronoi/Delaunay client paths.
- [x] Backend `pytest` covers the smooth and voronoi handlers and the updated dispatch-coverage guard.
- [x] Full build / typecheck and the existing frontend and backend suites pass.
- [x] Verified end-to-end in the running app with Playwright on real public data (52 US-state polygons, 1000 US-city points): grid produced the expected 12x5 = 60 cells from a bounding box, Smooth preserved all 52 features, Voronoi produced 1000 cells from 1000 points, and Delaunay produced 1990 triangles (consistent with 2n - 2 - h for an 8-vertex convex hull).

## Summary by CodeRabbit

* **New Features**
  * Added three vector processing tools: Smooth, Regular grid, and Voronoi/Delaunay (smoothing, grid generation, Voronoi/Delaunay from points).

* **Localization**
  * English labels added for the three new tools.

* **Tests**
  * Expanded unit and integration tests for smoothing, grid, Voronoi/Delaunay, 3D handling, parameter validation, and edge cases.

* **Chores**
  * CI matrix now includes Python 3.14; processing package gained Turf-based geometry dependencies.

### CODE REVIEW

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 1]
```diff
@@ -525,6 +525,167 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+def _smooth_ring(ring: list[list[float]], iterations: int) -> list[list[float]]:
+    closed = (
+        len(ring) > 1
+        and ring[0][0] == ring[-1][0]
+        and ring[0][1] == ring[-1][1]
+    )
+    pts = [[p[0], p[1]] for p in (ring[:-1] if closed else ring)]
+    for _ in range(iterations):
+        pts = _chaikin(pts, True)
+    pts.append([pts[0][0], pts[0][1]])
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py]: Done in b7afd4d — added a `if not pts: return pts` guard in `_smooth_ring` so an empty/degenerate ring yields an empty ring instead of an IndexError. The TypeScript mirror is guarded too.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 2]
```diff
@@ -525,6 +525,167 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+def _smooth(geojson, overlay, parameters) -> tuple[dict, list[str]]:
+    """Round the corners of line/polygon features with Chaikin's algorithm.
+
+    Works directly on the GeoJSON coordinates (no GeoPandas) so it produces
+    bit-identical output to the client engine, distinct from Simplify's vertex
+    reduction. Points pass through unchanged.
+    """
+    if not geojson or not geojson.get("features"):
+        raise ValueError("Input layer has no features")
+    iterations = int(parameters.get("iterations", 3) or 3)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py]: Done in b7afd4d — but using `math.floor(value + 0.5)` (with a non-finite fallback to 3) rather than `round()`. The client uses `Math.round`, which rounds half *up* (3.5 -> 4), whereas Python's `round()` is banker's rounding (3.5 -> 4 but 2.5 -> 2), so `round()` would not actually match. floor(x+0.5) reproduces Math.round for these positive values. Added a 3.5 -> 4 parity test.

File: packages/processing/src/vector-tools.ts [hunk 3]
```diff
@@ -1239,6 +1245,437 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+  const closed =
+    ring.length > 1 &&
+    ring[0][0] === ring[ring.length - 1][0] &&
+    ring[0][1] === ring[ring.length - 1][1];
+  let pts: Position[] = (closed ? ring.slice(0, -1) : ring).map((p) => [
+    p[0],
+    p[1],
+  ]);
+  for (let k = 0; k < iterations; k += 1) pts = chaikinOnce(pts, true);
+  pts.push([pts[0][0], pts[0][1]]);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts]: Done in b7afd4d — added `if (pts.length === 0) return pts;` before the closing push in `smoothRing`, mirroring the backend guard. Covered by a new empty-ring no-throw test.

File: packages/processing/src/vector-tools.ts [hunk 4]
```diff
@@ -1239,6 +1245,437 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+function chaikinOnce(points: Position[], closed: boolean): Position[] {
+  const n = points.length;
+  if (n < (closed ? 3 : 2)) return points;
+  const out: Position[] = [];
+  if (closed) {
+    for (let i = 0; i < n; i += 1) {
+      const a = points[i];
+      const b = points[(i + 1) % n];
+      out.push([a[0] * 0.75 + b[0] * 0.25, a[1] * 0.75 + b[1] * 0.25]);
+      out.push([a[0] * 0.25 + b[0] * 0.75, a[1] * 0.25 + b[1] * 0.75]);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts]: Done in b7afd4d — surfaced the limitation in `smoothTool.description` ("Z/elevation coordinates are not preserved."). Kept it 2D rather than interpolating Z so the client and Python engines stay bit-identical, which is the tool's stated invariant; Z interpolation can be added later if there's demand.

File: packages/processing/src/vector-tools.ts [hunk 5]
```diff
@@ -1239,6 +1245,437 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+    {
+      id: "cell_width",
+      label: "Cell width (degrees)",
+      type: "number",
+      required: true,
+      default: 1,
+      min: 0,
+      step: 0.1,
+    },
+    {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1500]: Done in b7afd4d — set `cell_width`/`cell_height` `min` to 0.0001 so the form can't submit 0, matching the runtime guard.

File: packages/processing/src/vector-tools.ts [hunk 6]
```diff
@@ -1239,6 +1245,437 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+              type: "Polygon",
+              coordinates: [
+                [
+                  [x0, y0],
+                  [x1, y0],
+                  [x1, y1],
+                  [x0, y1],
+                  [x0, y0],
+                ],
+              ],
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1628]: Respectfully disagreeing here — the ring is already counter-clockwise. With x1 > x0 and y1 > y0, the order [x0,y0] -> [x1,y0] -> [x1,y1] -> [x0,y1] is lower-left -> lower-right -> upper-right -> upper-left, i.e. CCW. Shoelace sum = (0) + (1) + (1) + (0) = +2 > 0, which is CCW under RFC 7946's right-hand rule. Applying the proposed reorder would make it clockwise, so leaving the winding as-is.

File: packages/processing/src/vector-tools.ts [hunk 7]
```diff
@@ -1239,6 +1245,440 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+      out.push([a[0] * 0.25 + b[0] * 0.75, a[1] * 0.25 + b[1] * 0.75]);
+    }
+    out.push([points[n - 1][0], points[n - 1][1]]);
+  }
+  return out;
+}
+
+/** Smooth an open line's coordinates with `iterations` Chaikin passes. */
+function smoothLine(coords: Position[], iterations: number): Position[] {
+  let pts: Position[] = coords.map((p) => [p[0], p[1]]);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts]: Done in 6ab5ccd — now preserving Z rather than dropping it: a shared `chaikinPoint`/`_chaikin_point` helper interpolates the third ordinate (`a[2]*wa + b[2]*wb`) when both endpoints are 3D, and endpoints are carried through with their Z. Applied identically in both engines so they stay bit-identical; added a 3D test (1/4 cut of z=100,200 -> 125). Description updated.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 8]
```diff
@@ -525,6 +525,181 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+        }
+    return geometry
+
+
+def _smooth(geojson, overlay, parameters) -> tuple[dict, list[str]]:
+    """Round the corners of line/polygon features with Chaikin's algorithm.
+
+    Works directly on the GeoJSON coordinates (no GeoPandas) so it produces
+    bit-identical output to the client engine, distinct from Simplify's vertex
+    reduction. Points pass through unchanged.
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py:649]: Already fixed in b7afd4d (before this round): the line is now `value = float(parameters.get("iterations", 3))` with a non-finite fallback to 3 and `iterations = math.floor(value + 0.5)` — no `or 3` truthiness, so an explicit 0 hits the range check and raises. Added test_smooth_iterations_zero_is_rejected to lock it.

File: packages/processing/src/vector-tools.ts [hunk 9]
```diff
@@ -1239,6 +1245,440 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+        ctx.log('Error: parameter "layer" has no GeoJSON features');
+        return;
+      }
+      bounds = bbox(layer.geojson) as [number, number, number, number];
+    }
+
+    const cellWidth = numberParam(ctx, "cell_width", NaN);
+    if (!Number.isFinite(cellWidth) || cellWidth <= 0) {
+      ctx.log("Error: cell width must be greater than 0");
+      return;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1571]: Done in 6ab5ccd — a non-positive/non-numeric cell_height now logs a Note before falling back to the cell width. A blank field still means "match width" and stays silent.

File: packages/processing/src/vector-tools.ts [hunk 10]
```diff
@@ -1239,6 +1245,440 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+    case "Polygon":
+      return {
+        type: "Polygon",
+        coordinates: geometry.coordinates.map((r) => smoothRing(r, iterations)),
+      };
+    case "MultiPolygon":
+      return {
+        type: "MultiPolygon",
+        coordinates: geometry.coordinates.map((poly) =>
+          poly.map((r) => smoothRing(r, iterations)),
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1346]: Done in 6ab5ccd — added a GeometryCollection case to smoothGeometry/_smooth_geometry that recurses over members (option 1), so line/polygon members are smoothed; such features are also counted in the log. Mirrored in both engines.

File: packages/processing/src/vector-tools.ts [hunk 11]
```diff
@@ -1239,6 +1245,440 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+    const cells = (result.features ?? []).filter((f) => Boolean(f?.geometry));
+    ctx.log(
+      `Voronoi: produced ${cells.length} cell(s) from ${points.length} point(s)`,
+    );
+    ctx.addResultLayer?.("Voronoi", featureCollection(cells));
+  },
+};
+
 export const VECTOR_TOOLS: ProcessingAlgorithm[] = [
   bufferTool,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1766]: Done in 6ab5ccd — added a zero-area-bbox guard (minX===maxX || minY===maxY) for both Voronoi and Delaunay on both engines, erroring with "points are collinear or coincident" instead of throwing or returning an empty result. Added a collinear test.

File: packages/processing/src/vector-tools.ts [hunk 12]
```diff
@@ -1239,6 +1245,440 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+        return;
+      }
+      bounds = [west, south, east, north];
+    } else {
+      const layer = getLayer(ctx, "layer");
+      if (!layer?.geojson?.features?.length) {
+        ctx.log('Error: parameter "layer" has no GeoJSON features');
+        return;
+      }
+      bounds = bbox(layer.geojson) as [number, number, number, number];
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1555]: Done in a1a162c — added a guard on the layer-extent path. Note `@turf/bbox` actually returns min/max (west <= east), so the literal antimeridian case can't arise there; the more real risk is a zero-area extent (e.g. a single-point layer, west === east), so the guard covers both `bounds[0] >= bounds[2] || bounds[1] >= bounds[3]` and errors clearly instead of logging a 0-cell grid.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 13]
```diff
@@ -525,6 +525,181 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+            "coordinates": [_smooth_ring(ring, iterations) for ring in coords],
+        }
+    if gtype == "MultiPolygon":
+        return {
+            "type": gtype,
+            "coordinates": [
+                [_smooth_ring(ring, iterations) for ring in poly] for poly in coords
+            ],
+        }
+    return geometry
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py:641]: Done in a1a162c — the four coordinate branches in `_smooth_geometry` now use `coords or []`, so a geometry missing its `coordinates` key yields an empty coordinate list instead of a TypeError (500). Added test_smooth_missing_coordinates_does_not_crash.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 14]
```diff
@@ -525,6 +525,181 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+        return _to_feature_collection(result), [message]
+    # Clip the (otherwise unbounded outer) cells to the points' bbox expanded by a
+    # 10% margin, matching the client, so they get a finite extent.
+    minx, miny, maxx, maxy = multipoint.bounds
+    dx = (maxx - minx) or 1
+    dy = (maxy - miny) or 1
+    envelope = box(minx - dx * 0.1, miny - dy * 0.1, maxx + dx * 0.1, maxy + dy * 0.1)
+    diagram = voronoi_diagram(multipoint, envelope=envelope)
+    cells = [cell.intersection(envelope) for cell in diagram.geoms]
+    cells = [cell for cell in cells if not cell.is_empty]
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py]: Done in a1a162c — the clip filter now also requires `geom_type in ("Polygon", "MultiPolygon")`, so an edge-coincident GeometryCollection (polygon + stray line) can't land in the output GeoDataFrame.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 15]
```diff
@@ -525,6 +525,181 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+    # non-finite/unparseable value falls back to 3, and math.floor(x + 0.5)
+    # rounds half up exactly like JS Math.round (not Python round()'s
+    # banker's rounding), keeping the two engines bit-identical.
+    try:
+        value = float(parameters.get("iterations", 3))
+    except (TypeError, ValueError):
+        value = 3.0
+    if not math.isfinite(value):
+        value = 3.0
+    iterations = math.floor(value + 0.5)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py]: Done in a1a162c — added an `isinstance(raw_iterations, bool)` check that falls back to 3 before the float() coercion, matching the client (numberParam returns the fallback for a non-number). Added test_smooth_boolean_iterations_falls_back_to_default.

File: packages/processing/src/vector-tools.ts [hunk 16]
```diff
@@ -1239,6 +1245,440 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+    ring[0][1] === ring[ring.length - 1][1];
+  let pts: Position[] = (closed ? ring.slice(0, -1) : ring).map((p) => [
+    p[0],
+    p[1],
+  ]);
+  for (let k = 0; k < iterations; k += 1) pts = chaikinOnce(pts, true);
+  // An invalid (empty/degenerate) ring leaves pts empty; guard so a malformed
+  // feature yields an empty ring rather than throwing on pts[0].
+  if (pts.length === 0) return pts;
+  pts.push([pts[0][0], pts[0][1]]);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts]: Done in a1a162c — both mirrors now `return []` when the de-closed ring has fewer than 3 vertices (was only the empty case), so 1-/2-vertex degenerate rings collapse to an empty ring instead of being re-closed into invalid GeoJSON. Added a degenerate-ring test on both engines.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 17]
```diff
@@ -525,6 +525,218 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+        return pts
+    pts.append(list(pts[0]))
+    return pts
+
+
+def _smooth_geometry(geometry: dict, iterations: int) -> dict:
+    """Smooth one GeoJSON geometry; non-line/polygon geometries pass through."""
+    gtype = geometry.get("type")
+    coords = geometry.get("coordinates")
+    if gtype == "LineString":
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py:611]: Already fixed in a1a162c (prior round): `_smooth_geometry` now uses `coords or []` in all four coordinate branches, so a geometry missing its `coordinates` key yields an empty list instead of a TypeError. Covered by test_smooth_missing_coordinates_does_not_crash.

File: packages/processing/src/vector-tools.ts [hunk 18]
```diff
@@ -1239,6 +1245,483 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+    // Both diagrams are undefined for collinear/coincident points (a zero-area
+    // bounding box). Turf's tin/voronoi would throw or return nothing; bail with
+    // a clear message instead. Mirrors the backend guard.
+    const [minX, minY, maxX, maxY] = bbox(pointsFc) as [
+      number,
+      number,
+      number,
+      number,
+    ];
+    if (minX === maxX || minY === maxY) {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1710]: Done in c671763 — the axis-aligned bbox guard stays, and now a 0-feature result from tin/voronoi_diagram (which is what diagonally collinear points like (0,0),(1,1),(2,2) produce) is reported as a collinear error on both engines rather than silently returning nothing. Added axis-aligned + diagonal Delaunay tests.

File: packages/processing/src/vector-tools.ts [hunk 19]
```diff
@@ -1239,6 +1245,483 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+        return;
+      }
+      bounds = [west, south, east, north];
+    } else {
+      const layer = getLayer(ctx, "layer");
+      if (!layer?.geojson?.features?.length) {
+        ctx.log('Error: parameter "layer" has no GeoJSON features');
+        return;
+      }
+      bounds = bbox(layer.geojson) as [number, number, number, number];
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1555]: Already fixed in a1a162c (prior round): the layer path now guards `bounds[0] >= bounds[2] || bounds[1] >= bounds[3]`, which covers a zero-area extent (coincident points -> [x,y,x,y]) and the all-null-geometry case (turf returns [Infinity, Infinity, -Infinity, -Infinity], so west >= east). Covered by the single-point-layer test.

File: packages/processing/src/vector-tools.ts [hunk 20]
```diff
@@ -1239,6 +1245,483 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+    if (!Number.isFinite(cellWidth) || cellWidth <= 0) {
+      ctx.log("Error: cell width must be greater than 0");
+      return;
+    }
+    const rawHeight = ctx.parameters.cell_height;
+    let cellHeight = numberParam(ctx, "cell_height", NaN);
+    if (!Number.isFinite(cellHeight) || cellHeight <= 0) {
+      // A blank field intentionally means "match the cell width"; only note it
+      // when the user actually entered an invalid (non-positive/non-numeric) value.
+      if (rawHeight !== undefined && rawHeight !== null && rawHeight !== "") {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts]: Nit 2: done in c671763 (`rawHeight != null && rawHeight !== ""`). Nit 1: declining — `detectGeometryProfile` (packages/map/src/geojson-loader.ts:34-43) already recurses into GeometryCollection and sets hasLine/hasPolygon for its members, so a GC layer containing lines/polygons does pass the ["line","polygon"] picker filter. The only excluded case is a points-only GeometryCollection, which Smooth would no-op anyway, so the filter is correct as-is.

File: tests/processing.test.ts [hunk 21]
```diff
@@ -743,6 +743,289 @@ describe("processing registry", () => {
...
+        },
+      ],
+      parameters: { layer: "collinear", type: "voronoi" },
+      log: (m) => collinearLogs.push(m),
+      addResultLayer: () => {
+        collinearProduced = true;
+      },
+    });
+    assert.equal(collinearProduced, false);
+    assert.ok(collinearLogs.some((m) => m.includes("collinear")));
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/processing.test.ts:1093]: Done in c671763 — added the Delaunay assertion alongside Voronoi in both suites (and a diagonal-collinear case), so a future refactor moving the guard below the diagram-type branch would fail a test.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 22]
```diff
@@ -525,6 +525,247 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+            "GeometryCollection",
+        ):
+            smoothed += 1
+        out_features.append(
+            {
+                "type": "Feature",
+                "properties": feature.get("properties") or {},
+                "geometry": _smooth_geometry(geometry, iterations),
+            }
+        )
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py]: Done in e7de69f — `_smooth` now preserves `id` (`if "id" in feature: new_feature["id"] = feature["id"]`), matching the null-geometry passthrough. Client mirrored. Added test_smooth_preserves_feature_id.

File: packages/processing/src/vector-tools.ts [hunk 23]
```diff
@@ -1239,6 +1245,509 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+      const isSmoothable =
+        isFamily(geometry, "line") ||
+        isFamily(geometry, "polygon") ||
+        geometry.type === "GeometryCollection";
+      if (isSmoothable) smoothed += 1;
+      return {
+        type: "Feature",
+        properties: feature.properties ?? {},
+        geometry: smoothGeometry(geometry, iterations),
+      };
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1418]: Done in e7de69f — applied the spread (`...(feature.id !== undefined ? { id: feature.id } : {})`) in the client and the equivalent in the Python backend; covered by a Feature.id test on both engines.

File: packages/processing/src/vector-tools.ts [hunk 24]
```diff
@@ -1239,6 +1245,509 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+    }
+    let smoothed = 0;
+    const features: Feature[] = fc.features.map((feature) => {
+      const geometry = feature.geometry;
+      if (!geometry) return feature;
+      const isSmoothable =
+        isFamily(geometry, "line") ||
+        isFamily(geometry, "polygon") ||
+        geometry.type === "GeometryCollection";
+      if (isSmoothable) smoothed += 1;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1410]: Done in e7de69f — isSmoothable now requires a GeometryCollection to contain a line/polygon member before counting it; a points-only collection no longer inflates the "Smoothed N" count. Mirrored in the backend.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 25]
```diff
@@ -525,6 +525,247 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
     return _to_feature_collection(result), [message]
 
 
+# Largest iteration count Smooth accepts; kept in sync with the client engine.
+_SMOOTH_MAX_ITERATIONS = 10
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py:529]: Declining to change this one: duplicating a constant across the TS and Python engines with a "kept in sync" comment is the established convention throughout this codebase (e.g. MAX_FEATURES, the AGGREGATE_STATS / spatial-join / select predicate sets), since there's no shared module across the two languages. The values are pinned by tests on each side (e.g. the iterations range check). Introducing cross-language enforcement here would be inconsistent with the rest of the codebase; happy to revisit if the maintainers want a project-wide approach.

File: .github/workflows/publish-python.yml [hunk 26]
```diff
@@ -28,7 +28,7 @@ jobs:
       matrix:
         # Cover the declared minimum (requires-python >= 3.11) and current
         # releases so newer-only syntax can't slip in unnoticed.
-        python-version: ["3.11", "3.12", "3.13"]
+        python-version: ["3.11", "3.12", "3.13", "3.14"]
```
[the remark this answers is not in the document]
Comment (giswqs) [reply .github/workflows/publish-python.yml:31]: This is on a separate commit by the maintainer (915f1b1, "Add support for Python 3.14") that landed on this branch, not part of the vector-tools change. Leaving it open for @giswqs to decide — the concern about 3.14 wheel availability for numpy/geopandas/shapely/pyarrow is worth a look, but it's outside this PR's scope.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 27]
```diff
@@ -525,6 +525,253 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+            value = 3.0
+    if not math.isfinite(value):
+        value = 3.0
+    iterations = math.floor(value + 0.5)
+    if iterations < 1 or iterations > _SMOOTH_MAX_ITERATIONS:
+        raise ValueError(f"Iterations must be between 1 and {_SMOOTH_MAX_ITERATIONS}")
+    out_features = []
+    smoothed = 0
+    _smoothable = ("LineString", "MultiLineString", "Polygon", "MultiPolygon")
+    for feature in geojson["features"]:
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py:675]: Done in 9c171b5 — `_smooth` now raises `ValueError("Each feature must be a GeoJSON Feature object")` for a null/non-dict feature before calling `.get()`, so the sidecar returns a clean 400 instead of a 500. Added test_smooth_null_feature_raises_value_error.

File: packages/processing/src/vector-tools.ts [hunk 28]
```diff
@@ -1239,6 +1245,517 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+        `Delaunay: produced ${result.features.length} triangle(s) from ${points.length} point(s)`,
+      );
+      ctx.addResultLayer?.("Delaunay", result);
+      return;
+    }
+    // Clip the Voronoi cells to the points' bounding box expanded by a 10% margin,
+    // matching the backend, so the outer (otherwise unbounded) cells get a finite
+    // extent rather than spanning the whole world.
+    const dx = maxX - minX;
+    const dy = maxY - minY;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1737]: Done in 9c171b5 — the client now filters Voronoi cells to `Polygon`/`MultiPolygon` (was `Boolean(f?.geometry)`), matching the backend so a degenerate non-polygon clip result can't reach the output. The existing Voronoi test already asserts every output cell is Polygon/MultiPolygon.

File: packages/processing/src/vector-tools.ts [hunk 29]
```diff
@@ -1239,6 +1245,517 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+      type: "select",
+      default: "polygon",
+      options: [
+        { value: "polygon", label: "Rectangles" },
+        { value: "point", label: "Points (cell centers)" },
+      ],
+    },
+  ],
+  run: (ctx) => {
+    const source = (ctx.parameters.source as string) || "viewport";
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1520]: Done in 9c171b5 — the viewport guard now checks `view[0] >= view[2] || view[1] >= view[3]`, symmetric with the layer and bbox paths (message generalized to "empty or crosses the antimeridian").

File: packages/processing/src/vector-tools.ts [hunk 30]
```diff
@@ -1239,6 +1245,517 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+        ctx.log("Error: map viewport is unavailable");
+        return;
+      }
+      if (view[0] >= view[2]) {
+        ctx.log(
+          "Error: the map view crosses the antimeridian; pan so it doesn't wrap +/-180, or use a manual bounding box",
+        );
+        return;
+      }
+      bounds = view;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts]: Already fixed in 9c171b5 (prior round): the viewport guard is now `if (view[0] >= view[2] || view[1] >= view[3])`, symmetric with the layer/bbox paths. This comment reflects the pre-9c171b5 code.

File: packages/processing/src/vector-tools.ts [hunk 31]
```diff
@@ -1239,6 +1245,517 @@ export const aggregateTool: ProcessingAlgorithm = {
...
+        points.push({
+          type: "Feature",
+          properties: feature.properties ?? {},
+          geometry: { type: "Point", coordinates },
+        });
+      }
+    }
+  }
+  return points;
+}
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/processing/src/vector-tools.ts:1659]: Intentional, and as you note it has no user-visible effect: Voronoi/Delaunay output cells carry no input id (they don't map back to seed features), and multiple sub-points of one MultiPoint would share the parent id anyway. Leaving collectPoints as-is to avoid another churn round; happy to add the clarifying comment if you'd like it in.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 32]
```diff
@@ -525,6 +525,258 @@ def _aggregate(geojson, overlay, parameters) -> tuple[dict, list[str]]:
...
+    diagram = voronoi_diagram(multipoint, envelope=envelope)
+    cells = [cell.intersection(envelope) for cell in diagram.geoms]
+    # Clipping a cell whose edge coincides with the envelope can yield a
+    # GeometryCollection (polygon + stray line); keep only polygonal parts so a
+    # non-Polygon geometry never lands in the output.
+    cells = [
+        cell
+        for cell in cells
+        if not cell.is_empty and cell.geom_type in ("Polygon", "MultiPolygon")
+    ]
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py:770]: Both are working as intended: (1) the two-step `cells` filter is left readable as you suggest; (2) counting after the polygon filter is deliberate — `produced N cell(s)` should report what's actually output, so N < M when boundary slivers are dropped is correct, not a bug. Leaving as-is rather than spawning another review round for a comment-only change.
